feat: Allow ASGs of empty size if matching a name.

dianakhuang opened the pull request:
In order to do certain types of cutover testing,
we would like to allow ASGs to be created with no instances in them if they match a certain name.

It doesn't seem like the clearing out of old ASGs goes through this flow as `_iterate_and_delete_empty_asgs` is only called if `_asg_is_empty` is true when bringing up a new asg.

doctoryes commented on tubular/asgard.py:
```diff
@@ -180,7 +180,12 @@ def new_asg(cluster, ami_id):
+    if response.status_code == 404:
+        msg = "Can't create more ASGs for cluster {}, please wait " \
+              "till older ASGs have been cleaned up or manually clean " \
+              "up old ASGs."
```
How about this text instead?:
Can't create more ASGs for cluster {}. Please either wait until older ASGs have been removed automatically or remove old ASGs manually via Asgard.

doctoryes commented on tubular/tests/test_asgard.py:
```diff
@@ -517,6 +517,23 @@ def post_callback(request, uri, headers):
+            httpretty.POST,
+            asgard.NEW_ASG_URL,
+            body=post_callback,
+            )
```
Nit: Move this paren back four spaces.

feanil commented:
@doctoryes ready for re-review.

doctoryes commented:
👍 👍 👍

dianakhuang committed:
feat: Allow ASGs of empty size if matching a name.

In order to do certain types of cutover testing,
we would like to allow ASGs to be created with no instances
in them if they match a certain name.

This skips the exception being thrown if the excluded name matches
the regular expression.

timmc-edx commented on tubular/asgard.py:
```diff
@@ -305,6 +310,10 @@ def _asg_is_empty(asg):
     Returns:
        bool: True if empty; False otherwise
     """
+
```
Short-circuiting here would make the function name and docstring incorrect -- and I think we should call that delete-empty cleanup function anyhow. (It doesn't delete the final ASG even if empty, and `delete_asg` itself refuses by default to delete the only remaining ASG in a cluster.)

What I would suggest instead is putting a conditional around the `raise ASGCountZeroException` and maybe defining a function for checking if an ASG is allowed to be empty. (Either way we should update the docstring of `new_asg`, which claims the resulting ASG will not be empty.)

I was a little concerned that there might be downstream issues with empty ASGs but it looks like even the "wait for healthy" check [should be fine](https://github.com/edx/tubular/blob/648f0008dd57d72244b5bb86250d6e7b092b817a/tubular/ec2.py#L511) with them. (There might be other issues though.)

timmc-edx commented on tubular/asgard.py:
```diff
@@ -52,6 +54,9 @@
 MAX_ATTEMPTS = int(os.environ.get('RETRY_MAX_ATTEMPTS', 5))
 
+# This is a list of ASGs we would like to allow to be empty.
+ALLOW_EMPTY = [r"stage-edx-WorkerServerASGroup*"]
```
Need a `.` here, and just for best practices there should be anchoring:

```suggestion
ALLOW_EMPTY = [r"^stage-edx-WorkerServerASGroup.*"]
```

timmc-edx commented on tubular/asgard.py:
```diff
@@ -305,6 +312,7 @@ def _asg_is_empty(asg):
     Returns:
        bool: True if empty; False otherwise
     """
+
```
[nit] Stray whitespace addition.

timmc-edx commented on tubular/asgard.py:
```diff
@@ -270,10 +275,12 @@ def new_asg(cluster, ami_id):
-                .format(asg_name=newest_asg['autoScalingGroupName'])
-        )
+        # For any ASGs where the name matches the regexes listed in ALLOW_EMPTY, don't raise an exception.
+        if not any([re.search(exclude_regex, newest_asg["autoScalingGroupName"]) for exclude_regex in ALLOW_EMPTY]):
```
Docstring of fn still needs update.

timmc-edx reviewed (approved):
A couple of small items, but otherwise looks good.
